aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -1423,53 +1455,77 @@ void q_batch_norm_kernel(
   AT_DISPATCH_QINT_TYPES(input.scalar_type(), "qbatch_norm", [&]() {
     float* alpha = a.data_ptr<float>();
     float* beta = b.data_ptr<float>();
+    auto minimum = std::numeric_limits<scalar_t::underlying>::lowest();
+    auto maximum = std::numeric_limits<scalar_t::underlying>::max();
     scalar_t::underlying* X =
         reinterpret_cast<scalar_t::underlying*>(input.data_ptr());
     scalar_t::underlying* Y = reinterpret_cast<scalar_t::underlying*>(output.data_ptr());
 
     constexpr int kVLen = 8;
     const int64_t outer_size = N * HxW;
     using Vec = Vec256<scalar_t>;
+
     // Hoisted variables
     auto in_zp_vec = Vec256<float>(static_cast<float>(in_zero_point));
     auto fake_scale = Vec256<float>(1.0f);
     auto scale_neg_zp_premul = fake_scale * in_zp_vec.neg();
     auto out_zero_point_v = Vec(scalar_t(out_zero_point));
-
-    // TODO replace with TensorIterator implementation once #33166 is fixed.
+    size_t lanes = Vec::float_num_vecs() * kVLen;
     for (int64_t i = 0; i < outer_size; ++i) {
-      int64_t n = C / (Vec::float_num_vecs() * kVLen) * (Vec::float_num_vecs() * kVLen);
-      int64_t r = C % (Vec::float_num_vecs() * kVLen);
       auto* X_ptr = reinterpret_cast<typename scalar_t::underlying*>(X + i * C);
       auto* Y_ptr = reinterpret_cast<typename scalar_t::underlying*>(Y + i * C);
-
-      for (int64_t j = 0; j < n; j += Vec::float_num_vecs() * kVLen) {
-        auto vals_q = Vec::loadu(X_ptr + j);
-        // Fake scale of 1.0 here, should not affect performance (FMA in place of sub)
-        auto vals_dq = vals_q.dequantize(fake_scale, in_zp_vec, scale_neg_zp_premul);
-        for (size_t idx = 0; idx < vals_dq.size(); ++idx) {
-          auto alpha_v = Vec256<float>::loadu(alpha + j + idx * kVLen);
-          auto beta_v = Vec256<float>::loadu(beta + j + idx * kVLen);
-          vals_dq[idx] = vec256::fmadd(alpha_v, vals_dq[idx], beta_v);
-        }
-        // Fake scale again
-        auto outputs_q = Vec::quantize(vals_dq, /*output_scale=*/1.0f, out_zero_point, /*inv_output_scale=*/1.0f);
-        if (ReluFused) {
-          outputs_q = outputs_q.relu(out_zero_point_v);
-        }
-        outputs_q.store(Y_ptr + j);
+      int64_t ch = 0;
+
+      for(; ch + lanes <= C; ch += lanes ) {
```

Comment:
Can we use `parallel_for` here for further speedup?

Reply:
Hmm, we did this in the batch dimension.